algebra/localization.lean

```diff
@@ -86,6 +97,169 @@ by refine
+     λ hnr₂, hy ⟨⟦⟨s₂, r₂, hnr₂⟩⟧, by rw ←hrs₂;
+       from quotient.sound ⟨1, is_submonoid.one_mem _, by simp [mul_comm]⟩⟩,
+     or.cases_on (mem_or_mem_of_mul_eq_zero P ht) hts $
+     λ h, have h1 : (r₁ * s₂ + r₂ * s₁) * r₃ - s₁ * s₂ * s₃ ∈ P,
+     by simpa using h,
+     have h2 : (r₁ * s₂ + r₂ * s₁) * r₃ ∈ P,
+     from is_ideal.mul_mem $ is_ideal.add_mem
+         (is_ideal.mul_mem hr₁)
+         (is_ideal.mul_mem hr₂),
+     or.cases_on (is_prime_ideal.mem_or_mem_of_mul_mem $
+         is_ideal.mem_of_mem_of_sub_mem h2 h1)
+       (λ h3, or.cases_on (is_prime_ideal.mem_or_mem_of_mul_mem h3) hs₁ hs₂)
+       hs₃)
+
+end at_prime
+
+def closure (S : set α) : set α := {y | ∃ (L:list α) (H:∀ x ∈ L, x ∈ S), L.prod = y }
+
+instance closure.is_submonoid (S : set α) : is_submonoid α (closure S) :=
+{ one_mem := ⟨[], by simp⟩,
+  mul_mem := λ x₁ x₂ ⟨L₁, hLS₁, hL₁⟩ ⟨L₂, hLS₂, hL₂⟩,
+    ⟨L₁ ++ L₂,
+     λ x hx, or.cases_on (list.mem_append.1 hx) (hLS₁ x) (hLS₂ x),
+     by simp [list.prod_append, *]⟩ }
+
+variable α
+
+def non_zero_divisors : set α := {x | ∀ z, x * z = 0 → z = 0}
+
+instance non_zero_divisors.is_submonoid : is_submonoid α (non_zero_divisors α) :=
+{ one_mem := λ z hz, by simpa using hz,
+  mul_mem := λ x₁ x₂ hx₁ hx₂ z hz,
+    have x₁ * (x₂ * z) = 0,
+    by rwa mul_assoc at hz,
+    hx₂ z $ hx₁ (x₂ * z) this }
+
+def quotient_ring := loc α (non_zero_divisors α)
+
+section quotient_ring
+
+variables {β : Type u} [integral_domain β] [decidable_eq β]
+
+lemma ne_zero_of_mem_non_zero_divisors {x : β} :
+  x ∈ loc.non_zero_divisors β → x ≠ 0 :=
+λ hm hz, have x * 1 = 0, by simp [hz], zero_ne_one (hm 1 this).symm
+
+lemma eq_zero_of_ne_zero_of_mul_eq_zero {x y : β} : x ≠ 0 → x * y = 0 → y = 0 :=
+λ hnx hxy, match eq_zero_or_eq_zero_of_mul_eq_zero hxy with
+| or.inl hx := false.elim $ hnx hx
+| or.inr hy := hy
+end
+
+lemma mem_non_zero_divisors_of_ne_zero {x : β} :
+  x ≠ 0 → x ∈ loc.non_zero_divisors β :=
+λ hnx z, eq_zero_of_ne_zero_of_mul_eq_zero hnx
+
+variable β
+
+def inv_aux : β × (non_zero_divisors β) → quotient_ring β :=
+λ ⟨r, s, hs⟩, if h : r = 0 then
+  ⟦⟨0, 1, is_submonoid.one_mem _⟩⟧
+else ⟦⟨s, r, mem_non_zero_divisors_of_ne_zero h⟩⟧
+
+def inv : quotient_ring β → quotient_ring β :=
+quotient.lift (inv_aux β) $
+λ ⟨r₁, s₁, hs₁⟩ ⟨r₂, s₂, hs₂⟩ ⟨t, hts, ht⟩,
+begin
+  have hrs : r₁ * s₂ - r₂ * s₁ = 0,
+  { from hts _ ht },
+  by_cases hr₁ : r₁ = 0;
+  by_cases hr₂ : r₂ = 0;
+  simp [inv_aux, hr₁, hr₂],
+  { exfalso,
+    simp [hr₁] at hrs,
+    exact ne_zero_of_mem_non_zero_divisors hs₁
+      (eq_zero_of_ne_zero_of_mul_eq_zero hr₂ hrs) },
+  { exfalso,
+    simp [hr₂] at hrs,
+    exact ne_zero_of_mem_non_zero_divisors hs₂
+      (eq_zero_of_ne_zero_of_mul_eq_zero hr₁ hrs) },
+  { exact ⟨1, is_submonoid.one_mem _,
+    by simpa [mul_comm] using congr_arg (λ x, -x) hrs⟩ }
+end
+
+instance quotient_ring.field.of_integral_domain : field (quotient_ring β) :=
+by refine
+{ loc.comm_ring β _ with
```

Comment:
use `..H` instead of `H with` notation (the latter is deprecated). But it shouldn't be needed here, assuming you already have an instance of `comm_ring (quotient_ring B)`.